go/vt/sysvars/sysvars.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2020 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sysvars
+
+// This information lives here, because it's needed from the vtgate planbuilder, the vtgate engine,
+// and the AST rewriter, that happens to live in sqlparser.
+
+// SystemVariable is a system variable that Vitess handles in queries such as:
+// select @@sql_mode
+// set skip_query_plan_cache = true
+type SystemVariable struct {
+	// IsBoolean is used to signal necessary type coercion so that strings
+	// and numbers can be evaluated to a boolean value
+	IsBoolean bool
+
+	// IdentifierAsString allows identifiers (a.k.a. ColName) from the AST to be handled as if they are strings.
+	// SET transaction_mode = two_pc => SET transaction_mode = 'two_pc'
+	IdentifierAsString bool
+
+	// Default is the default value, if none is given
+	Default string
+
+	Name string
+}
+
+var (
+	on          = "1"
+	off         = "0"
+	multi       = "'MULTI'"
+	unspecified = "'UNSPECIFIED'"
+	utf8        = "'utf8'"
+
+	Autocommit          = SystemVariable{Name: "autocommit", IsBoolean: true, Default: on}
+	ClientFoundRows     = SystemVariable{Name: "client_found_rows", IsBoolean: true, Default: off}
+	SkipQueryPlanCache  = SystemVariable{Name: "skip_query_plan_cache", IsBoolean: true, Default: off}
+	TxReadOnly          = SystemVariable{Name: "tx_read_only", IsBoolean: true, Default: off}
+	TransactionReadOnly = SystemVariable{Name: "transaction_read_only", IsBoolean: true, Default: off}
+	SQLSelectLimit      = SystemVariable{Name: "sql_select_limit", Default: off}
+	TransactionMode     = SystemVariable{Name: "transaction_mode", Default: multi, IdentifierAsString: true}
+	Workload            = SystemVariable{Name: "workload", Default: unspecified, IdentifierAsString: true}
+	Charset             = SystemVariable{Name: "charset", Default: utf8, IdentifierAsString: true}
+	Names               = SystemVariable{Name: "names", Default: utf8, IdentifierAsString: true}
```

Comment:
Defining defaults like this is dangerous. most of them are flags, so they can be different from the default on start of vtgate. Why do we need default when those are available on session

Reply:
Added a commit that changes this behaviour after our talk